### ISSUE TITLE
re-architecture of pipenv formula burden

### ISSUE THREAD

**Issue by kennethreitz** [Homebrew/homebrew-core#22429]
**tl;dr**: *Pipenv wants to own its own formula, similar to how Heroku appears to.*

----------

Pipenv is now the officially recommended packaging tool for Python (from [Python.org](https://packaging.python.org/tutorials/managing-dependencies/)), and we'd love for homebrew to be the offically recommended way to install the tool.

Unfortunately, today, this is not the case, as the formula does not rely on the Python3 formula and instead depends on the system Python (which is not actually a valid/kosher CPython interpreter, but a mostly–compatible fork thereof). 

-------------

I see in my personal installation two forumla: `heroku` and `heroku/brew/heroku`. I'm under the impression that this means that Heroku (my company, actually) is under the control of their own formula and is responsible for updating it and the maintenance burden has been removed from the homebrew core team. 

---------------

I haven't asked anyone at my company yet, but is this an accurate picture of the way this is implemented? If so, I think pipenv would heavily benefit from the same setup. 

It would allow us to forumlate the forumla as appropriate, not burden the maintainers of homebrew with unneccessary nitpicks moving forward, and in general allow us to recommend homebrew as the de-facto best practice installation method moving forward — which is something I think we both want. 

Thoughts?

/cc @erinxocon

**Comment by ilovezfs** [comment]
Um https://github.com/Homebrew/homebrew-core/blob/master/Formula/pipenv.rb#L17

**Comment by kennethreitz** [comment]
Sorry, I was a little out of date on the python3 issue. 

Still, I think owning our own forumla might be best, and that this issue deserves to be kept open until discussion about that occurs.

**Comment by ilovezfs** [comment]
>Still, I think owning our own forumla might be best, and that this issue deserves to be kept open until discussion about that occurs.

That is not how Homebrew works. There are no formula "owners." Anyone is welcome to open a pull request for any formula, and only the Homebrew maintainers are allowed to merge them or give final approvals :-1: or :-1:.

**Comment by ilovezfs** [comment]
Regarding heroku specifically, you can talk to @jdxcode and see https://github.com/Homebrew/homebrew-core/pull/15319 for how that formula is handled.

**Comment by ilovezfs** [comment]
@kennethreitz that said, in practice there are some upstreams who are consistently the only people that open PRs against their formula and take very good care of them. For example, see
https://github.com/Homebrew/homebrew-core/pulls?q=is%3Apr+ponyc+is%3Aclosed

>It would allow us to forumlate the forumla as appropriate, not burden the maintainers of homebrew with unneccessary nitpicks moving forward, and in general allow us to recommend homebrew as the de-facto best practice installation method moving forward — which is something I think we both want.

Whether you want to recommend Homebrew in this way is up to you, of course. I don't think you'll have any problems doing so, but of course I would say that, right? https://github.com/Homebrew/homebrew-core/graphs/contributors

**Comment by kennethreitz** [comment]
No need to be aggressive, I was just trying to make things easier for everyone, as we often make multiple releases a day when in a burst :)

**Comment by ilovezfs** [comment]
lol. I'm not :heart: Sorry if it seemed that way …

**Comment by kennethreitz** [comment]
No worries — sounds like all is well with the current situation :)

**Comment by ilovezfs** [comment]
@kennethreitz so one question is whether you want us to keep the formula [as it is now](https://github.com/Homebrew/homebrew-core/blob/8a01ac6a2e9cbeb72e3cb7b792fae9c01e89b062/Formula/pipenv.rb#L19-L77) with a dependency tree that we update, and pin, at each release, which allows the builds to be semi-reproducible (modulo the exact python3 version and Homebrew/brew commit) or make it dynamic like awscli is https://github.com/Homebrew/homebrew-core/blob/3a6b7416a770fdca92bc73de1d5ff7c6827708d7/Formula/awscli.rb#L23-L27 which can be easier maintain but is theoretically more subject to random breakage due to unexpected changes in the dependencies.

**Comment by kennethreitz** [comment]
I think the awscli approach would be wise.

**Comment by kennethreitz** [comment]
does awscli lose “brew upgrade” capabilities w/ point releases then?

**Comment by ilovezfs** [comment]
@kennethreitz I've throttled awscli brew upgrades to every 10 upstream releases because they have a new one every five minutes, but that's unrelated to the lack of dep pinning.
